doc/developer/design/20240925_network_policies.md

```diff
@@ -0,0 +1,148 @@
+- Preventing out-of-policy traffic from reaching Environmentd.
+ (note this means network policies will not prevent DDOS)
+- Policy inheritance from associated roles. IE if 'bob' is a member of role 'eng'
+we will not apply policies from role 'eng' to 'bob'.
+- Restricting global API access.
+- Restricting access to Frontegg.
+
+## Solution Proposal
+
+#### Overview
+The proposed solution is to use role-based policies with a default network policy that applies to any role without a policy. This can initially be implemented as a global default policy and will be extended to per-user and per-source/sink. The policy will be applied when an attempt is made to establish a new client connection with the coordinator.
+
+#### New Resources
+A new `NetworkPolicy` resource will be added to the catalog.
+```rust
+struct NetworkPolicy {
+    id: NetworkPolicyId
+    name: String,
+    rules: Vec<NetworkPolicyRule>,
+}
+
+enum NetworkPolicyRule {
+    Ingress {
+        action: NetworkPolicyRuleAction,
+        source: IpNet,
+        comment: String
+ }
+}
+
+enum NetworkPolicyRuleAction {
+    Allow
+    // Deny - may be added later
+}
+```
+
+Users will be able to create `NetworkPolicies` directly.  A user must have `CREATENETWORKPOLICY` privileges to create, modify, or destroy network policies. Network policies will be limited to 25 rules. This will be controlled by an LD flag. `NetworkPolicyRules` must be created through a policy. The policy rules implementation will initially only contain an `Allow` variant, but we should be an enum to allow for a `Deny` variant in the future. `NetworkPolicyRule` will be an enum to allow for both ingress and egress policies. Only ingress policies will be initially implemented. `NetworkPolicyRules::Ingress` will also contain a single `IpNet` and a comment text field. Comments have become a standard for rules and greatly increase the manageability and auditability of policies.
+
+Example syntax for creating a network policy
+```sql
+CREATE NETWORK POLICY OFFICE_01 (
+ RULE ( ACTION=ALLOW, SOURCE="10.0.0.0/32", COMMENT="OFFICE IP - 2024-9-28" )
+);
+```
+
+Network policies will be assignable to roles - eventually source/sinks as well. A user must have usage privileges for the network policy they wish to assign, as well as the privileges to modify the role.
+
+Example syntax for assigning a network policy to a role
+```sql
+ALTER ROLE BOB SET network_policy = OFFICE_01;
+```
+* Policies can only be applied to login roles. There is no policy inheritance. Only the policy assigned to the role the user logged in as will be checked.
+
+Along with network policies a new `SystemVar` (`default_network_policy`) will be added that points to a specific `NetworkPolicy`. This system var will only be modifiable by `mz_system` and `superuser`. If a resource does not have a network policy this policy will be applied.
+
+Example syntax for updating the default_network_policy
+```sql
+ALTER SYSTEM SET default_network_policy = OFFICE_01;
```

Comment:
The way our variable inheritance works I think you just want to call this `network_policy`; variables that aren't overridden at the role level default to the system parameter of the same name. This is how e.g. `cluster` works.

Reply:
ok, I think I'm not fully understanding this. 

My understanding for clusters is that there is a system var for cluster that selects which cluster will be selected by default on login. Users can update this var to select a different cluster. I think we have a different story for policies. Customers shouldn't be able to set their policy through a session var. In order to apply a different policy to a role you'd have to alter the role and set a new policy, or alter the default policy for everyone. 

I'm happy to change the name to `network_policy` just want to make sure we're agreeing on the flow.

Reply:
> Customers shouldn't be able to set their policy through a session var. In order to apply a different policy to a role you'd have to alter the role and set a new policy, or alter the default policy for everyone.

💯 agree with this.

The reason I think it's still ideal to use `network_policy` for both the system and role variable is because you're going to need to add special cases to deny inheritance as a session variable. By default, anything settable as a role variable (e.g. `cluster`) is also settable as a session variable. If you add a system variable named `default_network_policy` and a role variable named `network_policy`, you're going to need to add special cases in a few places to handle the inheritance properly. Whereas if you just use `network_policy`, the existing inheritance should work correctly for the system/role level, and you'll only need to add one special case to ensure that users can't override the `network_policy` at the session level.